Nodejs/Product/Nodejs/SharedProject/FileChangeManager.cs

```diff
@@ -227,27 +277,54 @@ internal void IgnoreItemChanges(string fileName, bool ignore)
         /// Stop observing when the file is updated on disk.
         /// </summary>
         /// <param name="fileName">File to stop observing.</param>
-        internal void StopObservingItem(string fileName)
+        public void StopObservingFile(string fileName)
         {
+            this.CheckDisposed();
+
             if (string.IsNullOrEmpty(fileName))
             {
                 throw new ArgumentException(SR.GetString(SR.InvalidParameter), nameof(fileName));
             }
 
             var fullFileName = Utilities.CanonicalizeFileName(fileName);
 
-            if (this.observedItems.ContainsKey(fullFileName))
+            // Remove the file from our observed list. It's important that this is done before the call to 
+            // UnadviseFileChange, because for some reason, the call to UnadviseFileChange can trigger a 
+            // FilesChanged event, and we want to be able to filter that event away.
+            if (this.observedFiles.TryRemove(fullFileName, out var value))
             {
                 // Get the cookie that was used for this.observedItems to this file.
-                var itemInfo = this.observedItems[fullFileName];
+                var (ItemID, FileChangeCookie) = value;
 
-                // Remove the file from our observed list. It's important that this is done before the call to 
-                // UnadviseFileChange, because for some reason, the call to UnadviseFileChange can trigger a 
-                // FilesChanged event, and we want to be able to filter that event away.
-                this.observedItems.Remove(fullFileName);
+                // Stop observing the file
+                ErrorHandler.ThrowOnFailure(this.fileChangeService.UnadviseFileChange(FileChangeCookie));
+            }
+        }
 
+        public bool StopObservingFolder(string folderName)
+        {
+            this.CheckDisposed();
+
+            if (string.IsNullOrEmpty(folderName))
+            {
+                throw new ArgumentException(SR.GetString(SR.InvalidParameter), nameof(folderName));
+            }
+
+            var fullFolderName = Utilities.CanonicalizeFileName(folderName);
+
+            if (this.observedFolders.TryRemove(fullFolderName, out var cookie))
+            {
                 // Stop observing the file
-                ErrorHandler.ThrowOnFailure(this.fileChangeService.UnadviseFileChange(itemInfo.FileChangeCookie));
+                return ErrorHandler.Succeeded(this.fileChangeService.UnadviseDirChange(cookie));
+            }
+            return false;
+        }
+
+        private void CheckDisposed()
```

Comment:
>CheckDisposed [](start = 21, length = 13)

Does this make sense in a class accessed from multiple threads?  It looks like it could be disposed after performing the check but before executing the actual logic.